feature-push-notifications/src/main/java/io/novafoundation/nova/feature_push_notifications/data/presentation/handling/types/TokenReceivedNotificationHandler.kt

```diff
@@ -0,0 +1,106 @@
+    private val context: Context,
+    private val accountRepository: AccountRepository,
+    private val tokenRepository: TokenRepository,
+    private val deepLinkConfigurator: DeepLinkConfigurator<AssetDetailsLinkConfigPayload>,
+    override val chainRegistry: ChainRegistry,
+    notificationIdProvider: NotificationIdProvider,
+    gson: Gson,
+    notificationManager: NotificationManagerCompat,
+    resourceManager: ResourceManager,
+) : BaseNotificationHandler(
+    notificationIdProvider,
+    gson,
+    notificationManager,
+    resourceManager,
+    channel = NovaNotificationChannel.TRANSACTIONS
+),
+    PushChainRegestryHolder {
+
+    override suspend fun handleNotificationInternal(channelId: String, message: RemoteMessage): Boolean {
+        val content = message.getMessageContent()
+        content.requireType(NotificationTypes.TOKENS_RECEIVED)
+        val chain = content.getChain()
+        val recipient = content.extractPayloadField<String>("recipient")
+        val assetId = content.extractPayloadField<String?>("assetId")
+        val amount = content.extractBigInteger("amount")
+
+        val metaAccountsQuantity = accountRepository.getActiveMetaAccountsQuantity()
+        val recipientMetaAccount = accountRepository.findMetaAccount(chain.accountIdOf(recipient), chain.id) ?: return false
+
+        val notification = NotificationCompat.Builder(context, channelId)
+            .buildWithDefaults(
+                context,
+                getTitle(metaAccountsQuantity, recipientMetaAccount),
+                getMessage(chain, assetId, amount),
+                makeAssetDetailsPendingIntent(deepLinkConfigurator, chain.id, chain.utilityAsset.id)
+            ).build()
+
+        notify(notification)
+
+        return true
+    }
+
+    private fun getTitle(metaAccountsQuantity: Int, senderMetaAccount: MetaAccount?): String {
+        val accountName = senderMetaAccount?.formattedAccountName()
+        return when {
+            metaAccountsQuantity > 1 && accountName != null -> resourceManager.getString(R.string.push_token_received_title, accountName)
+            else -> resourceManager.getString(R.string.push_token_received_no_account_name_title)
+        }
+    }
+
+    private suspend fun getMessage(
+        chain: Chain,
+        assetId: String?,
+        amount: BigInteger
+    ): String {
+        val asset = chain.assets.firstOrNull { it.onChainAssetId == assetId } ?: chain.utilityAsset
+        val token = tokenRepository.getTokenOrNull(asset)
+        val tokenAmount = amount.formatPlanks(asset)
+        val fiatAmount = token?.planksToFiat(amount)
+            ?.formatAsCurrency(token.currency)
+
+        return when {
+            fiatAmount != null -> resourceManager.getString(R.string.push_token_received_message, tokenAmount, fiatAmount, chain.name)
+            else -> resourceManager.getString(R.string.push_token_received_message_no_fiat, tokenAmount, chain.name)
+        }
```

Comment:
I do not entirely support such duplication coverage; I think that in case of changes, these cases should be modified separately, because there is no guarantee that the changes will be applied equally to both handlers.

But I agree that some specific logic should be taken out separately like this line: `chain.assets.firstOrNull { it.onChainAssetId == assetId } ?: chain.utilityAsset`

Reply:
I think that hard fixing this duplication is not a good idea since we also have a different number of format string parameters for send and for receive

Reply:
I guess the main concern for me he is the dublication of "calculate fiat and do something when we it is present and do something else otherwise"
In another case it is the same but with meta accounts quantity

Reply:
Fixed